rust/lance/src/dataset/udtf.rs

```diff
@@ -0,0 +1,352 @@
+///                     "terms": "catch fish",
+///                     "operator": "And"
+///                 }
+///             }
+///      "#;
+/// let options = r#"
+///             {
+///                 "with_row_id": true
+///             }
+///      "#;
+/// let df = ctx
+///  .sql(&format!("SELECT * FROM fts('HarryPotter_Chapter1', '{}', '{}') WHERE number > 1",
+///        fts_query, options))
+///  .await
+///  .unwrap();
+/// ```
+#[derive(Debug)]
+struct FtsQueryUDTF {
+    datasets: HashMap<String, Arc<Dataset>>,
+}
+
+impl TableFunctionImpl for FtsQueryUDTF {
+    fn call(&self, expr: &[Expr]) -> datafusion::common::Result<Arc<dyn TableProvider>> {
+        // Parse params: table_name, fts_query, options
+        if expr.len() < 2 || expr.len() > 3 {
+            return Err(DataFusionError::Execution(
+                "FtsQueryUDTF function takes table_name, fts_query and optional options as parameters".to_string(),
+            ));
+        }
+
+        let Some(Expr::Literal(ScalarValue::Utf8(Some(table_name)), _)) = expr.first() else {
+            return Err(DataFusionError::Execution(
+                "FtsQueryUDTF first argument should be table name in string".to_string(),
+            ));
+        };
+
+        let Some(Expr::Literal(ScalarValue::Utf8(Some(fts_query)), _)) = expr.get(1) else {
+            return Err(DataFusionError::Execution(
+                "FtsQueryUDTF second argument should be fts query in json format".to_string(),
+            ));
+        };
+
+        let (with_row_id, with_row_addr, ordered) =
+            if let Some(Expr::Literal(ScalarValue::Utf8(Some(options)), _)) = expr.get(2) {
+                parse_query_options(options)?
+            } else {
+                (false, false, false)
+            };
+
+        // Fts query
+        let dataset = self
+            .datasets
+            .get(table_name)
+            .ok_or_else(|| DataFusionError::Execution(format!("Table {} not found", table_name)))?;
+
+        let provider = FtsTableProvider::new(
+            dataset.clone(),
+            FullTextSearchQuery::new_query(from_json(fts_query)?),
+            with_row_id,
+            with_row_addr,
+            ordered,
+        );
+        Ok(Arc::new(provider))
+    }
+}
+
+fn parse_query_options(options: &str) -> datafusion::common::Result<(bool, bool, bool)> {
+    let value: Value = serde_json::from_str(options)
+        .map_err(|e| Error::invalid_input(format!("invalid json options: {}", e), location!()))?;
+    let with_row_id = value
+        .get("with_row_id")
+        .is_some_and(|v| v.as_bool().unwrap_or(false));
+    let with_row_addr = value
+        .get("with_row_addr")
+        .is_some_and(|v| v.as_bool().unwrap_or(false));
+    let ordered = value
+        .get("ordered")
+        .is_some_and(|v| v.as_bool().unwrap_or(false));
+    Ok((with_row_id, with_row_addr, ordered))
+}
+
+/// Builder of `FtsQueryUDTF`
+#[allow(dead_code)]
+struct FtsQueryUDTFBuilder {
```

Comment:
I think we should make this struct and the module public so people can use it directly?